src/xc_integrator/replicated/device/incore_replicated_xc_device_integrator_exc_vxc.hpp

```diff
@@ -116,8 +116,97 @@ void IncoreReplicatedXCDeviceIntegrator<ValueType>::
+  auto rt  = detail::as_device_runtime(this->load_balancer_->runtime());
+  auto device_data_ptr = lwd->create_device_data(rt);
+
+  GAUXC_MPI_CODE( MPI_Barrier(rt.comm());) 
+
+  // Temporary electron count to judge integrator accuracy
+  value_type N_EL;
+
+  if( this->reduction_driver_->takes_device_memory() ) {
+
+    // If we can do reductions on the device (e.g. NCCL)
+    // Don't communicate data back to the host before reduction
+    this->timer_.time_op("XCIntegrator.LocalWork_EXC_VXC", [&](){
+      exc_vxc_local_work_( basis, Ps, ldps, Pz, ldpz, nullptr, 0, nullptr, 0, tasks.begin(), tasks.end(), 
+        *device_data_ptr);
+    });
+
+    GAUXC_MPI_CODE(
+    this->timer_.time_op("XCIntegrator.ImbalanceWait_EXC_VXC",[&](){
+      MPI_Barrier(this->load_balancer_->runtime().comm());
+    });  
+    )
+
+    // Reduce results in device memory
+    auto vxc_s_device = device_data_ptr->vxc_s_device_data();
+    auto vxc_z_device = device_data_ptr->vxc_z_device_data();
+    auto exc_device = device_data_ptr->exc_device_data();
+    auto nel_device = device_data_ptr->nel_device_data();
+    auto queue      = device_data_ptr->queue();
+    this->timer_.time_op("XCIntegrator.Allreduce_EXC_VXC", [&](){
+      this->reduction_driver_->allreduce_inplace( vxc_s_device, nbf*nbf, ReductionOp::Sum, queue );
+      this->reduction_driver_->allreduce_inplace( vxc_z_device, nbf*nbf, ReductionOp::Sum, queue );
+      this->reduction_driver_->allreduce_inplace( exc_device, 1,       ReductionOp::Sum, queue );
+      this->reduction_driver_->allreduce_inplace( nel_device, 1,       ReductionOp::Sum, queue );
+    });
+
+    // Retrieve data to host
+    this->timer_.time_op("XCIntegrator.DeviceToHostCopy_EXC_VXC",[&](){
+      device_data_ptr->retrieve_exc_vxc_integrands( EXC, &N_EL, VXCs, ldvxcs, VXCz, ldvxcz );
+    });
+
+
+  } else {
+
+    // Compute local contributions to EXC/VXC and retrieve
+    // data from device 
+    this->timer_.time_op("XCIntegrator.LocalWork_EXC_VXC", [&](){
+      exc_vxc_local_work_( basis, Ps, ldps, Pz, ldpz, nullptr, 0, nullptr, 0,
+                                VXCs, ldvxcs, VXCz, ldvxcz, nullptr, 0, nullptr, 0, EXC, 
+                              &N_EL, tasks.begin(), tasks.end(), *device_data_ptr);
+    });
+
+    GAUXC_MPI_CODE(
+    this->timer_.time_op("XCIntegrator.ImbalanceWait_EXC_VXC",[&](){
+      MPI_Barrier(this->load_balancer_->runtime().comm());
+    });  
+    )
+
+    // Reduce Results in host mem
+    this->timer_.time_op("XCIntegrator.Allreduce_EXC_VXC", [&](){
+      this->reduction_driver_->allreduce_inplace( VXCs, nbf*nbf, ReductionOp::Sum );
+      this->reduction_driver_->allreduce_inplace( VXCz, nbf*nbf, ReductionOp::Sum );
+      this->reduction_driver_->allreduce_inplace( EXC,   1    , ReductionOp::Sum );
+      this->reduction_driver_->allreduce_inplace( &N_EL, 1    , ReductionOp::Sum );
+    });
+
+  }
 }
 
 template <typename ValueType>
```

Comment:
From an implementation stand point, this function should be made nullptr safe (as the local_work already is) and the RKS/UKS drivers should just delegate to it. Otherwise there's a ton of code replication